Label masking example

Padarn opened the pull request:
This PR adds an example of label masking using the UniMP architecture from https://arxiv.org/abs/2009.03509. 

Addresses https://github.com/pyg-team/pytorch_geometric/issues/4040

- [x] Create a data loader that creates masks for training with masked labels rather than just manually doing this

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,130 @@
+train_mask = torch.zeros(data.x.size(0))
+train_mask = train_mask.scatter_(0, split_idx["train"], 1).type(torch.bool)
+
+valid_mask = torch.zeros(data.x.size(0))
```
```suggestion
valid_mask = torch.zeros(data.num_nodes, dtype=torch.bool)
```

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,130 @@
+train_mask = train_mask.scatter_(0, split_idx["train"], 1).type(torch.bool)
+
+valid_mask = torch.zeros(data.x.size(0))
+valid_mask = valid_mask.scatter_(0, split_idx["valid"], 1).type(torch.bool)
```
I prefer
```suggestion
valid_mask[split_idx['valid']] = True
```
since I found it to be a bit easier to read.

Padarn replied:
yep this is nicer cheers!

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,130 @@
+
+
+class UnimpNet(torch.nn.Module):
+    def __init__(self, feature_size: int, label_vocab: int, inner_dim: int,
```
```suggestion
    def __init__(self, feature_size: int, num_classes: int, inner_dim: int,
```

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,130 @@
+        label_mask: torch.Tensor,
+    ):
+        x = torch.clone(x)
+        x[~label_mask] = x[~label_mask] + self.label_embedding(
```
```suggestion
        x[~label_mask] += self.label_embedding(
```
should work as well, right?

Padarn replied:
makes sense... I was actually trying to avoid the clone which felt like a bit of a hack (to be honest I didn't fully understand what was going on here, but it seemed like x was being free'd after the backwards pass because I'm modifying it in place).

rusty1s replied:
I don't think we can avoid the `clone` here since it looks like PyTorch does not expose a `masked_add` function :(

Padarn replied:
I changed it to use a scatter and a zero vector.. but it might be worse 😓

rusty1s reviewed (commented):
I think this looks great. Thank you!

I like your suggestion in https://github.com/pyg-team/pytorch_geometric/issues/4040#issuecomment-1059890976 to add a pre-process layer to PyG, e.g., like
```python
class MaskLabel(torch.nn.Module):
    def __init__(num_classes, in_channels):
        self.emb = ...

    def forward(self, x, y, mask):
```
that could also provide utility for generating the masks. Generating masks seems hard to automate though, as we would need to distinguish between the different splits and which labels we are allowed to make use of.

Padarn commented:
> Generating masks seems hard to automate though

Yeah its a bit tricky. Probably the best option is to just provide some basic utility like 

```
def ratio_mask(mask, ratio):
     # sets len(mask == 1)*ratio elements to zero
```

And then provide examples (like the one in this PR of how to use it.

I'll have a play around with trying to make the utility and use it in the example to see if it feels usable.

rusty1s commented:
FYI: https://github.com/pyg-team/pytorch_geometric/pull/4201

Padarn commented:
Oh cool, will incorporate. 

I updated the example with some examples we discussed (we can move these into the main library once we're aligned on the basics)

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,170 @@
+        self.out_channels = out_channels
+
+    def forward(self, y: torch.Tensor, mask: torch.Tensor):
+        out = torch.zeros(y.shape[0], self.out_channels, dtype=torch.float)
```
I think we should expect both `x` and `y` as input arguments (just like done previously). This lets us avoid the intermediate `out` creation and saves some memory.

Padarn replied:
Hmm, can, but this means we need to do a clone.. the zero avoid this need because then we can write the non-in-place
```
x = x + label_embedding(y)
```

So we'd have to clone x here. What do you think is better?

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,170 @@
+        model.train()
+
+        # create epoc training mask that chooses subset of train to remove
+        epoc_mask = ratio_mask(train_mask, 1 - label_rate)
```
```suggestion
        epoch_mask = ratio_mask(train_mask, 1 - label_rate)
```

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,170 @@
+        shuffle (bool): Whether or not the mask is pre-shuffled, if so there is
+            no need to randomize which entires are set to :obj:`False`.
+    """
+    n = int(mask.count_nonzero().item())
```
```suggestion
    n = int(mask.sum())
```

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,170 @@
+            no need to randomize which entires are set to :obj:`False`.
+    """
+    n = int(mask.count_nonzero().item())
+    new_mask = torch.ones(len(mask), dtype=torch.bool)
```
can't we initialize `new_mask` based on `mask`? Since we are doing `mask & new_mask` later anyway.

Padarn replied:
good point - do you think clone is the cleanest way to do this?

rusty1s commented on examples/unimp_arxiv.py:
```diff
@@ -0,0 +1,170 @@
+heads = 2
+
+
+def ratio_mask(mask: torch.Tensor, ratio: float, shuffle: bool = False):
```
I like it. How about we add `ratio_mask` as a (static) method to `MaskLabel`?

Padarn replied:
💯

Padarn commented on torch_geometric/nn/models/mask_label.py:
```diff
@@ -0,0 +1,48 @@
+        self.emb = torch.nn.Embedding(num_classes, out_channels)
+        self.out_channels = out_channels
+
+    def forward(self, y: torch.Tensor, mask: torch.Tensor):
```
@rusty1s - I didn't change this yet. My feeling is that if we provide `x` then we 1) need to clone it, 2) force the way in which `y` is combined with `x` (sum vs concat for example). 

What do you think? Maybe these limitations are fine and its more usable providing `x`?

rusty1s replied:
I think we should change it. The `copy` isn't a problem at all and will actually lead to a more efficient solution. We can also think about providing options of how a user wants to combine both `x` and `y`. Otherwise, the functionality of `MaskLabel` is kinda limited :)

Padarn replied:
Yeah okay fair enough. Perhaps we should just allow for sum and concat then?

rusty1s replied:
Sounds good :)

Padarn replied:
Updated. By the way just to confirm my understanding - the copy is more efficient because eventually, it copies anyway when the out-of-place operation happens? Intuitively copy is more expensive then creating zero tensor so I'm missing something.

rusty1s replied:
I think it should be more efficient since the copy approach requires only one access to the entire feature matrix (the copy), while the summation approach requires 2 (both for `x` and `y`).

Padarn replied:
Ah that makes sense. Thank you!

Padarn commented on torch_geometric/nn/models/mask_label.py:
```diff
@@ -0,0 +1,48 @@
+                to change to :obj:`False`.
+        """
+        n = int(mask.sum())
+        mask = torch.clone(mask)
```
Not sure this is much better than before in terms of efficiency, but its more readable, and now deterministic as to how many elements are masked

rusty1s replied:
I'm not sure I understand the solution. Can't we simply do `mask[mask == True] = torch.rand(n) > ratio`?

Padarn replied:
Can, but then it maybe less than `n*ratio` that are set to zero. Also presumably we need to clone mask otherwise we'll in place modify the training mask?

rusty1s replied:
Understood, thanks for clarifying. Let's keep it that way then.

Padarn commented:
Hey @rusty1s I've added some tests and refactored the new layer into the main library. Left a few comments on some final choices, happy to take your final advice on these.

rusty1s commented on torch_geometric/nn/models/__init__.py:
```diff
@@ -20,39 +20,15 @@
+    'ARGA', 'ARGVA', 'SignedGCN', 'RENet', 'GraphUNet', 'SchNet', 'DimeNet',
+    'Explainer', 'to_captum', 'GNNExplainer', 'MetaPath2Vec', 'DeepGCNLayer',
+    'TGNMemory', 'LabelPropagation', 'CorrectAndSmooth', 'AttentiveFP',
+    'RECT_L', 'LINKX', 'LightGCN', 'MaskLabel'
```
```suggestion
    'RECT_L', 'LINKX', 'LightGCN', 'MaskLabel',
```

rusty1s commented:
Thanks for the updates. Did you run the example? Is the model performance reasonable at the moment?

Padarn commented:
Well its a bit slow on my laptop :-) But yes the model seems to converge towards reasonable performance. Quite a bit lower than the leaderboard, but seemed to get in the 70% range which seems typical.

Padarn commented:
```
INFO:/workspaces/pytorch_geometric/examples/unimp_arxiv.py:
            epoch = 74:
            train loss = 1.110485315322876, train acc = 71.45%
            valid loss = 1.2410309314727783, valid acc = 67.00%
            test loss = 1.1324297189712524, test acc = 69.37%
```

The performance is not great, but its in the right ballpark. Any suggestions for other parameters to try?

rusty1s commented:
We could try to replicate the model hyperparameters of the official implementation. It currently looks like we are only making use of a single layer (without any non-linearity).

Padarn commented:
okay makes sense 👍

Padarn commented:
Updated to be very close to the papers setup, but I couldn't run with hidden dimension 128, even with the same GPU they say they use I ran out of memory.

```
INFO:examples/unimp_arxiv.py:
            epoch = 499:
            train loss = 0.8647370934486389, train acc = 74.05%
            valid loss = 0.8711884021759033, valid acc = 73.07%
            test loss = 0.8791332244873047, test acc = 73.22%
```

rusty1s commented:
Pretty cool! Thanks for the update. I can try to look into the memory issue.

Padarn commented:
Sorry I just realized my GPU has half the memory of the one used in the reference paper.. so it makes sense. If you have a bigger GPU you can give it a go.

emalgorithm commented on torch_geometric/nn/models/mask_label.py:
```diff
@@ -0,0 +1,59 @@
+    Classification" <https://arxiv.org/abs/2009.03509>`_ paper.
+
+    In the forward pass both the labels, and a mask corresponding to which
+    labels should be kept is provided. All entires that are not true in the
```
Typo: "entires" -> "entries"

emalgorithm commented on torch_geometric/nn/models/mask_label.py:
```diff
@@ -0,0 +1,59 @@
+            return torch.cat([x, out], dim=1)
+        else:
+            x = torch.clone(x)
+            x[mask] = self.emb(y[mask])
```
This should be adding rather than assigning I think

Padarn replied:
wow good catch, thats an embarassing one! will fix it soon

Padarn commented:
thanks a lot for your review @emalgorithm, I've fixed those.

@rusty1s thoughts on merging this?

rusty1s commented:
Wanted to merge it last weekend but didn't get to it. Sorry for the delay. Will do it ASAP.

Padarn commented:
No worries, glad you didn't, got some good fixes in

On Wed, Mar 30, 2022 at 2:09 PM Matthias Fey ***@***.***>
wrote:

> Wanted to merge it last weekend but didn't get to it. Sorry for the delay.
> Will do it ASAP.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pyg-team/pytorch_geometric/pull/4197#issuecomment-1082660247>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGRPN2BRBG3HLHKECWDJL3VCPVYTANCNFSM5P7M6MHQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

-- 


By communicating with Grab Inc and/or its subsidiaries, associate 
companies and jointly controlled entities (“Grab Group”), you are deemed to 
have consented to the processing of your personal data as set out in the 
Privacy Notice which can be viewed at https://grab.com/privacy/ 
<https://grab.com/privacy/>


This email contains confidential information 
and is only for the intended recipient(s). If you are not the intended 
recipient(s), please do not disseminate, distribute or copy this email 
Please notify Grab Group immediately if you have received this by mistake 
and delete this email from your system. Email transmission cannot be 
guaranteed to be secure or error-free as any information therein could be 
intercepted, corrupted, lost, destroyed, delayed or incomplete, or contain 
viruses. Grab Group do not accept liability for any errors or omissions in 
the contents of this email arises as a result of email transmission. All 
intellectual property rights in this email and attachments therein shall 
remain vested in Grab Group, unless otherwise provided by law.

Padarn commented:
Thanks! I'm curious too. Could obviously use a sampler but they don't in
the paper

On Thu, 10 Mar 2022, 10:51 pm Matthias Fey, ***@***.***>
wrote:

> Pretty cool! Thanks for the update. I can try to look into the memory
> issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pyg-team/pytorch_geometric/pull/4197#issuecomment-1064145220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGRPN46DDMJC2JC3CXYPPLU7ID6TANCNFSM5P7M6MHQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

-- 


By communicating with Grab Inc and/or its subsidiaries, associate 
companies and jointly controlled entities (“Grab Group”), you are deemed to 
have consented to the processing of your personal data as set out in the 
Privacy Notice which can be viewed at https://grab.com/privacy/ 
<https://grab.com/privacy/>


This email contains confidential information 
and is only for the intended recipient(s). If you are not the intended 
recipient(s), please do not disseminate, distribute or copy this email 
Please notify Grab Group immediately if you have received this by mistake 
and delete this email from your system. Email transmission cannot be 
guaranteed to be secure or error-free as any information therein could be 
intercepted, corrupted, lost, destroyed, delayed or incomplete, or contain 
viruses. Grab Group do not accept liability for any errors or omissions in 
the contents of this email arises as a result of email transmission. All 
intellectual property rights in this email and attachments therein shall 
remain vested in Grab Group, unless otherwise provided by law.
